e2e_playwright/empty_charts_test.py

```diff
@@ -0,0 +1,48 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import re
+
+from playwright.sync_api import Page, expect
+
+from e2e_playwright.conftest import ImageCompareFunction
+
+
+def test_gracefully_handles_no_data(app: Page, assert_snapshot: ImageCompareFunction):
+    charts = app.get_by_test_id("stArrowVegaLiteChart")
+    expect(charts).to_have_count(4)
+
+    for i in range(4):
+        assert_snapshot(
+            app.get_by_test_id("stArrowVegaLiteChart").nth(i),
+            name=f"st_vega_lite_chart_empty-{i}",
+        )
+
+    # pyplot graph assertion
+    assert app.get_by_test_id("stImage").locator("img").get_attribute("src") is not None
+
+
+def test_handles_no_data_with_exception(app: Page):
+    exception_messages = app.get_by_test_id("stException")
+    expect(exception_messages).to_have_count(5)
+
+    for i in range(4):
+        expect(exception_messages.nth(i)).to_have_text(
+            re.compile("ValueError: Vega-Lite charts require a non-empty spec dict.")
+        )
+
+    assert (
```

Comment:
Breaks rule 1 and 3 of: https://github.com/streamlit/streamlit/wiki/Running-e2e-tests-and-updating-snapshots#three-rules-of-playwright

Reply:
I think the issue with what you wrote in the PR description might be that some other chars here also have e meaning in the regex. What might be easier is just to check something like:
re.compile("TypeError: ArrowAltairMixin.altair_chart.*")

or maybe even better
```
except(exception_messages.nth(4)). to_contain_text("TypeError: ArrowAltairMixin.altair_chart() missing 1 required positional argument: 'altair_chart'")
```
or
```
except(exception_messages.nth(4)). to_contain_text("TypeError: ArrowAltairMixin.altair_chart")
```

Reply:
Thanks. Will check this out. Tho I don't recall seeing a to_contain_text api?

Reply:
Yeah, I also didn't discover this for some time. But it is very useful 👍